File: scripts/validate-contributor.js
#!/usr/bin/env node

/**
 * Validates all contributor JSON files
 */

const fs = require('fs');
const path = require('path');

function validateContributorFile(filepath) {
    const filename = path.basename(filepath);
    const username = filename.replace('.json', '');
    
    // Skip template files
    if (filename === 'example-contributor.json' || filename === '.gitkeep' || filename === 'README.md') {
        return { valid: true, skipped: true };
    }
    
    let data;
    try {
        const content = fs.readFileSync(filepath, 'utf8');
        data = JSON.parse(content);
    } catch (e) {
        return {
            valid: false,
            errors: [`Invalid JSON: ${e.message}`]
        };
    }
    
    const errors = [];
    const warnings = [];
    
    // Check required fields
    const required = ['name', 'github', 'contributions'];
    required.forEach(field => {
        if (!data[field]) {
            errors.push(`Missing required field: ${field}`);
        }
    });
    
    // Check github username matches filename
    if (data.github && data.github !== username) {
        errors.push(`GitHub username '${data.github}' doesn't match filename '${filename}'`);
    }
    
    // Check contributions is an array
    if (data.contributions && !Array.isArray(data.contributions)) {
        errors.push('contributions must be an array');
    } else if (data.contributions && data.contributions.length === 0) {
        errors.push('contributions array cannot be empty');
    }
    
    // Validate contribution types
    const validContributions = [
        'a11y', 'audio', 'blog', 'bug', 'business', 'code', 'content', 'data',
        'design', 'doc', 'eventOrganizing', 'example', 'financial', 'fundingFinding',
        'ideas', 'infra', 'maintenance', 'mentoring', 'platform', 'plugin',
        'projectManagement', 'promotion', 'question', 'research', 'review',
        'security', 'talk', 'test', 'tool', 'translation', 'tutorial',
        'userTesting', 'video'
    ];
    
    if (data.contributions && Array.isArray(data.contributions)) {
        data.contributions.forEach(contrib => {
            if (!validContributions.includes(contrib)) {
                warnings.push(`Unknown contribution type: ${contrib}`);
            }
        });
    }
    
    // Check profile URL format
    if (data.profile && !data.profile.startsWith('http')) {
        warnings.push('profile should be a full URL starting with http:// or https://');
    }
    
    return {
        valid: errors.length === 0,
        errors,
        warnings,
        data
    };
}

function main() {
    console.log('🔍 Validating contributor files...\n');
    
    const contributorsDir = path.join(process.cwd(), 'contributors');
    
    if (!fs.existsSync(contributorsDir)) {
        console.log('❌ Contributors directory not found!');
        process.exit(1);
    }
    
    const files = fs.readdirSync(contributorsDir).filter(f => f.endsWith('.json'));
    
    if (files.length === 0) {
        console.log('❌ No contributor JSON files found!');
        process.exit(1);
    }
    
    let totalFiles = 0;
    let validFiles = 0;
    let skippedFiles = 0;
    let errorFiles = 0;
    let totalWarnings = 0;
    
    files.forEach(file => {
        const filepath = path.join(contributorsDir, file);
        const result = validateContributorFile(filepath);
        
        if (result.skipped) {
            skippedFiles++;
            return;
        }
        
        totalFiles++;
        
        if (result.valid && result.warnings.length === 0) {
            console.log(`✅ ${file}`);
            validFiles++;
        } else if (result.valid && result.warnings.length > 0) {
            console.log(`⚠️  ${file}`);
            result.warnings.forEach(warning => {
                console.log(`   ⚠️  ${warning}`);
            });
            validFiles++;
            totalWarnings += result.warnings.length;
        } else {
            console.log(`❌ ${file}`);
            result.errors.forEach(error => {
                console.log(`   ❌ ${error}`);
            });
            if (result.warnings) {
                result.warnings.forEach(warning => {
                    console.log(`   ⚠️  ${warning}`);
                });
                totalWarnings += result.warnings.length;
            }
            errorFiles++;
        }
    });
    
    console.log('\n📊 Validation Summary:');
    console.log(`  Files checked: ${totalFiles}`);
    console.log(`  Valid: ${validFiles}`);
    console.log(`  Errors: ${errorFiles}`);
    console.log(`  Warnings: ${totalWarnings}`);
    if (skippedFiles > 0) {
        console.log(`  Skipped: ${skippedFiles} (template files)`);
    }
    
    if (errorFiles > 0) {
        console.log('\n❌ Validation failed! Fix the errors above.');
        process.exit(1);
    } else {
        console.log('\n✅ All contributor files are valid!');
    }
}

if (require.main === module) {
    main();
}

File: scripts/validate-contributor.py
#!/usr/bin/env python3
"""
Validation script for contributor JSON files
This script validates that contributor JSON files follow the expected format
"""

import json
import os
import sys
import re
from pathlib import Path

# Valid contribution types based on all-contributors specification
VALID_CONTRIBUTIONS = {
    "a11y", "audio", "blog", "bug", "business", "code", "content", "data",
    "design", "doc", "eventOrganizing", "example", "financial", "fundingFinding",
    "ideas", "infra", "maintenance", "mentoring", "platform", "plugin",
    "projectManagement", "promotion", "question", "research", "review",
    "security", "talk", "test", "tool", "translation", "tutorial", "userTesting",
    "video"
}

def validate_json_file(filepath):
    """Validate a single contributor JSON file"""
    errors = []
    warnings = []
    
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        return [f"Invalid JSON: {e}"], []
    except Exception as e:
        return [f"Error reading file: {e}"], []
    
    # Check required fields
    required_fields = ['name', 'github', 'contributions']
    for field in required_fields:
        if field not in data:
            errors.append(f"Missing required field: {field}")
        elif not data[field]:
            errors.append(f"Empty required field: {field}")
    
    # Validate name
    if 'name' in data:
        if not isinstance(data['name'], str):
            errors.append("Field 'name' must be a string")
        elif len(data['name'].strip()) == 0:
            errors.append("Field 'name' cannot be empty")
    
    # Validate github username
    if 'github' in data:
        github = data['github']
        if not isinstance(github, str):
            errors.append("Field 'github' must be a string")
        elif not re.match(r'^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$', github):
            errors.append(f"Invalid GitHub username format: {github}")
        else:
            # Check if filename matches GitHub username
            expected_filename = f"{github}.json"
            actual_filename = os.path.basename(filepath)
            if actual_filename != expected_filename:
                warnings.append(f"Filename '{actual_filename}' doesn't match GitHub username '{github}' (should be '{expected_filename}')")
    
    # Validate contributions
    if 'contributions' in data:
        contributions = data['contributions']
        if not isinstance(contributions, list):
            errors.append("Field 'contributions' must be an array")
        elif len(contributions) == 0:
            warnings.append("Field 'contributions' is empty")
        else:
            for contrib in contributions:
                if not isinstance(contrib, str):
                    errors.append(f"Contribution type must be a string: {contrib}")
                elif contrib not in VALID_CONTRIBUTIONS:
                    warnings.append(f"Unknown contribution type: {contrib}")
    
    # Validate optional profile field
    if 'profile' in data:
        profile = data['profile']
        if not isinstance(profile, str):
            errors.append("Field 'profile' must be a string")
        elif profile and not re.match(r'^https?://', profile):
            warnings.append(f"Profile URL should start with http:// or https://: {profile}")
    
    # Check for unexpected fields
    expected_fields = {'name', 'github', 'contributions', 'profile'}
    extra_fields = set(data.keys()) - expected_fields
    if extra_fields:
        warnings.append(f"Unexpected fields: {', '.join(extra_fields)}")
    
    return errors, warnings

def main():
    """Main validation function"""
    contributors_dir = Path("contributors")
    
    if not contributors_dir.exists():
        print("❌ Contributors directory not found")
        sys.exit(1)
    
    json_files = list(contributors_dir.glob("*.json"))
    if not json_files:
        print("❌ No JSON files found in contributors directory")
        sys.exit(1)
    
    total_errors = 0
    total_warnings = 0
    
    print(f"🔍 Validating {len(json_files)} contributor files...\n")
    
    for filepath in sorted(json_files):
        errors, warnings = validate_json_file(filepath)
        
        if errors or warnings:
            print(f"📁 {filepath.name}:")
            
            for error in errors:
                print(f"  ❌ {error}")
                total_errors += 1
            
            for warning in warnings:
                print(f"  ⚠️  {warning}")
                total_warnings += 1
            
            print()
        else:
            print(f"✅ {filepath.name}")
    
    print(f"\n📊 Validation Summary:")
    print(f"  Files checked: {len(json_files)}")
    print(f"  Errors: {total_errors}")
    print(f"  Warnings: {total_warnings}")
    
    if total_errors > 0:
        print("\n❌ Validation failed! Please fix the errors above.")
        sys.exit(1)
    elif total_warnings > 0:
        print("\n⚠️  Validation passed with warnings. Consider fixing the warnings above.")
        sys.exit(0)
    else:
        print("\n✅ All files are valid!")
        sys.exit(0)

if __name__ == "__main__":
    main()